### PR TITLE
Added remove_node/s methods and modified the original data structure

### PR DESCRIPTION
**Pull request by npochhi**
### Description

Implements remove_edge and remove_edges methods and makes relevant changes in the data structure.

- Relevant Issues : #8 #9 
- Relevant PRs : 
- Type of change :
  - [x] New feature
  - [ ] Bug fix for existing feature
  - [ ] Code quality improvement
  - [ ] Addition or Improvement of tests
  - [ ] Addition or Improvement of documentation

### CODE REVIEW

File: lib/networkx/graph.rb [hunk 1]
```diff
@@ -1,16 +1,24 @@
...
-      @nodes = Set.new
-      @edges = {}
+    attr_reader :_adj, :_nodes
+
+    def initialize(**graph_attrs)
+      @_nodes = {}
+      @_adj = {}
+      @graph = {}
+
+      @graph.merge!(graph_attrs)
```
Comment (athityakumar) [inline lib/networkx/graph.rb]:
```ruby
@graph = graph_attrs
```

File: lib/networkx/graph.rb [hunk 2]
```diff
@@ -1,16 +1,24 @@
...
-    def add_edge(node_1, node_2, **attrs)
+    def add_edge(node_1, node_2, **edge_attrs)
       add_node(node_1)
       add_node(node_2)
 
-      edge_key         = Set.new([node_1, node_2])
-      @edges[edge_key] = attrs
+      attrs_hash = @_adj[node_1][node_2]
+      attrs_hash = {} if attrs_hash.nil?
+      attrs_hash.merge!(edge_attrs)
```
Comment (athityakumar) [inline lib/networkx/graph.rb]: These 3 lines can be truncated into one line.

```ruby
edge_attrs = (@adj[node_1][node_2] || {}).merge(edge_attrs)
```

File: lib/networkx/graph.rb [hunk 3]
```diff
@@ -23,14 +31,37 @@ def add_edges(edges)
       end
     end
 
-    def add_node(node)
-      @nodes.add(node)
+    def add_node(node, **node_attrs)
+      if !@_nodes.key?(node)
```
Comment (athityakumar) [inline lib/networkx/graph.rb]: In general, the `if !x` block can be written as `unless x` for more readability. But here, can't  the `!` be removed by changing the order of checking? Something like the below.

```ruby
if @nodes.key?(node)
  @nodes[node].merge!(node_attrs)
else
  @adj[node] = {}
  @nodes[node] = node_attrs
end
```

File: lib/networkx/graph.rb [hunk 4]
```diff
@@ -23,14 +31,37 @@ def add_edges(edges)
...
+      end
+    end
+
+    def remove_node(node)
+      @_adj[node].each_key { |k| @_adj[k].delete(node) }
+      @_adj.delete(node)
+      @_nodes.delete(node)
+    rescue KeyError, NoMethodError
+      raise KeyError, "Error in deleting node #{node} from Graph."
+    end
```
Comment (athityakumar) [inline lib/networkx/graph.rb]: It'd be better to have a `@nodes.key?(node)` check rather than rely on a rescue block here.

**Comment by athityakumar** [comment]
Hey @npochhi. I've suggested a few changes and nit-pickings, do have a look. Also, try to add the YARD Doc comments and tests (in `spec/networkx/graph_spec.rb`) in this PR itself. 😄

**Comment by npochhi** [comment]
@athityakumar, should I add more tests?

**Comment by npochhi** [comment]
@athityakumar I've done the required changes. Please look at it.

**Comment by npochhi** [comment]
The green tick looks beautiful. :smile:

**Comment by athityakumar** [comment]
Haha, indeed. Meanwhile, it'd be better if the yard doc comments can be upgraded to modern specs (again, have a look at daru-io) and examples can be added. Documentation with just params description and no example, is going to be a pain in the neck for users to read through. 😄

File: spec/network_x/graph_spec.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
+require 'spec_helper'
```
Comment (athityakumar) [inline spec/network_x/graph_spec.rb]: Appropriately rename this to `spec/network_x/classes/graph_spec.rb`

**Comment by npochhi** [comment]
Is there no way to put `graph_spec` inside `classes` folder? Should I bypass this?

**Comment by athityakumar** [comment]
@npochhi - There are 3 options.

(1) Revert`lib/classes/graph.rb` to `lib/graph.rb` (I think this would actually be better, as all of our files will essentially be and for classes)
(2) Rename `spec/network_x/classes/graph_spec.rb` to something like `spec/network_x/classes/graph_class_spec.rb` (IMHO, `classes` directory doesn't really seem that worthy an inclusion)
(3) Ignore the `RSpec/FilePath` cop

**Comment by npochhi** [comment]
Done :smile:..

**Comment by MridulS** [comment]
@npochhi looks good :)
probably not a part of this PR but a quick question: How do you plan to handle node/edge (attributes) updates? Do you want to create a method to expose it like `update_node` or users should update the inner data structure, or maybe some other way.

**Comment by npochhi** [comment]
@MridulS I wouldn't like exposing the data structures as some erratic changes would lead to inconsistencies. I think a method like `update_node` or `update_edge` would be a better bet.

**Comment by npochhi** [comment]
@athityakumar We can do that. The reason NetworkX is not having a backend in lower level language is the usage of numpy for computation heavy algorithms. As numpy has a lower level backend, it need for a C/C++ backend reduces in case of NetworkX. Probably, we can take a thing from this and instead of writing the entire backend in C/C++, we can use nmatrix or any related library for our purpose.

**Comment by athityakumar** [comment]
@npochhi - As suggested in the video-call, would you like to add support to NMatrix backend for faster computation as a separate PR, or continue in this PR itself?

**Comment by npochhi** [comment]
I think it is better to implement the `to_nmatrix` method in other file, given that we'd have to add support for other libraries too. Hence, I think I'll like to implement the said method in another PR.

**Comment by athityakumar** [comment]
Merging this in. 😄

**Pull request merged**